cpp/src/io/json/experimental/read_json.cpp

```diff
@@ -38,29 +39,47 @@ size_t sources_size(host_span<std::unique_ptr<datasource>> const sources,
   });
 }
 
-std::vector<uint8_t> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
-                                      compression_type compression,
-                                      size_t range_offset,
-                                      size_t range_size)
+rmm::device_uvector<char> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
+                                           compression_type compression,
+                                           size_t range_offset,
+                                           size_t range_size,
+                                           rmm::cuda_stream_view stream)
 {
   CUDF_FUNC_RANGE();
   // Iterate through the user defined sources and read the contents into the local buffer
   auto const total_source_size = sources_size(sources, range_offset, range_size);
-  auto buffer                  = std::vector<uint8_t>(total_source_size);
-
-  size_t bytes_read = 0;
-  for (const auto& source : sources) {
-    if (!source->is_empty()) {
-      auto data_size   = (range_size != 0) ? range_size : source->size();
-      auto destination = buffer.data() + bytes_read;
-      bytes_read += source->host_read(range_offset, data_size, destination);
-    }
-  }
 
   if (compression == compression_type::NONE) {
-    return buffer;
+    auto d_buffer     = rmm::device_uvector<char>(total_source_size, stream);
+    size_t bytes_read = 0;
+    for (const auto& source : sources) {
+      if (!source->is_empty()) {
+        auto data_size   = (range_size != 0) ? range_size : source->size();
+        auto destination = reinterpret_cast<uint8_t*>(d_buffer.data()) + bytes_read;
+        if (source->is_device_read_preferred(data_size)) {
+          bytes_read += source->device_read(range_offset, data_size, destination, stream);
+        } else {
+          auto const h_buffer = source->host_read(range_offset, data_size);
+          CUDF_CUDA_TRY(cudaMemcpyAsync(destination,
+                                        h_buffer->data(),
+                                        h_buffer->size(),
+                                        cudaMemcpyHostToDevice,
+                                        stream.value()));
+          stream.synchronize();
```

Comment:
I believe this synchronization is needed to prevent the destruction of the host buffer before the copy has completed. That said, I wonder about `cudaMemcpyAsync` here as well.

Reply:
Yes, we sync because of the short life span of `h_buffer`. We could avoid syncing in each iteration if we keep the host buffer around for a bit longer, but I'm not sure it that would have a measurable impact. Let me know if you think this is a worthwhile change.
I default to `cudaMemcpyAsync` because it takes a stream, and we seem to be completely avoiding `cudaMemcpy` in libcudf (presumably because of the use of default stream?).

Reply:
Good answer. Thanks!